src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs

```diff
@@ -0,0 +1,314 @@
+        }
+
+        private InstrumentState? GetInstrumentState(Instrument instrument)
+        {
+            if (!_instrumentStates.TryGetValue(instrument, out InstrumentState? instrumentState))
+            {
+                lock (this) // protect _isntrumentConfigFuncs list
+                {
+                    foreach (var filter in _instrumentConfigFuncs)
+                    {
+                        if (filter(instrument))
+                        {
+                            instrumentState = BuildInstrumentState(instrument);
+                            if (instrumentState != null)
+                            {
+                                _instrumentStates.TryAdd(instrument, instrumentState);
+                                instrumentState = _instrumentStates[instrument];
+                            }
+                            break;
+                        }
+                    }
+                }
+            }
+            return instrumentState;
+        }
+
+        internal InstrumentState? BuildInstrumentState(Instrument instrument)
+        {
+            Func<Aggregator>? createAggregatorFunc = GetAggregatorFactory(instrument);
+            if (createAggregatorFunc == null)
+            {
+                return null;
+            }
+            Type aggregatorType = createAggregatorFunc.GetType().GenericTypeArguments[0];
+            Type instrumentStateType = typeof(InstrumentState<>).MakeGenericType(aggregatorType);
+            return (InstrumentState)Activator.CreateInstance(instrumentStateType, createAggregatorFunc)!;
+        }
+
+        private static Func<Aggregator>? GetAggregatorFactory(Instrument instrument)
+        {
+            Type type = instrument.GetType();
+            Type? genericDefType = null;
+#if IS_GENERIC_TYPE_SUPPORT
+            genericDefType = type.IsGenericType ? type.GetGenericTypeDefinition() : null;
+#else
+            try
+            {
+                 genericDefType = type.GetGenericTypeDefinition();
+            }
+            catch {}
+#endif
+            if (genericDefType == typeof(Counter<>))
+            {
+                return () => new RateSumAggregator();
+            }
+            else if (genericDefType == typeof(ObservableCounter<>))
+            {
+                return () => new RateAggregator();
+            }
+            else if (genericDefType == typeof(ObservableGauge<>))
+            {
+                return () => new LastValue();
+            }
+            else if (genericDefType == typeof(Histogram<>))
+            {
+                return () => new ExponentialHistogramAggregator(DefaultHistogramConfig);
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        internal void Collect()
+        {
+            _listener.RecordObservableInstruments();
+
+            foreach (KeyValuePair<Instrument, InstrumentState> kv in _instrumentStates)
```

Comment:
that's cool : )

Reply:
Not worth it if you need to write a method for it. I thought it would work for free due to [`KeyValuePair<Tk,Tv>.Deconstruct(out Tk key, out Tv value)`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.keyvaluepair-2.deconstruct?view=net-5.0).

Reply:
This library builds for some older .NET Framework versions where that API wasn't yet available.